File: go/diamond/diamond.go
package diamond

import (
	"errors"
	"strings"
)

func Gen(char byte) (string, error) {
	if char < 'A' || char > 'Z' {
		return "", errors.New("out of range A-Z")
	}

	var (
		b    byte
		c    byte = 'A'
		size      = int(char - c + 1)
		rows      = make([]string, 2*size-1)
	)

	for i := 0; i < size; i, c = i+1, c+1 {
		row := make([]byte, 2*size-1)
		for j := 0; j < size; j++ {
			b = ' '
			if j == i {
				b = c
			}
			row[(size-1)-j] = b
			row[(size-1)+j] = b
		}
		rows[i] = string(row)
		rows[2*(size-1)-i] = rows[i]
	}

	return strings.Join(rows, "\n"), nil

}

File: dart/diamond/lib/diamond.dart
extension on String {
  String get reversed => split('').reversed.join('');
  String skip(int n) => (length <= n) ? '' : substring(n);
}

class Diamond {
  final a = 'A'.codeUnitAt(0);

  List<String> rows(String letter) {
    int size = letter.codeUnitAt(0) - a + 1;
    if (size < 1 || size > 26) throw new ArgumentError("Invalid letter");

    /* a list of strings holding the right-hand half of the rows
     * for the top half of the diamond.
     */
    var topRightQuadrant = List.generate(size, (i) {
      var row = List.filled(size, ' ');
      row[i] = String.fromCharCode(a + i);
      return row.join('');
    });

    /* complete the left-hand half of the rows 
     */
    var topHalf =
        topRightQuadrant.map((row) => row.skip(1).reversed + row).toList();

    return topHalf + topHalf.reversed.skip(1).toList();
  }
}
